quadruped_pympc/controllers/Controllers.py

```diff
@@ -0,0 +1,543 @@
+import os
+import time
+
+# TODO: Ugly hack so people dont have to run the python command specifying the working directory.
+#  we should remove this before the final release.
+import sys
+
+# Add the parent directory of this script to the Python path
+sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+
+import mujoco
+import numpy as np
+
+
+# Parameters for both MPC and simulation
+from quadruped_pympc.helpers.foothold_reference_generator import FootholdReferenceGenerator
+from quadruped_pympc.helpers.periodic_gait_generator import PeriodicGaitGenerator
+from quadruped_pympc.helpers.swing_trajectory_controller import SwingTrajectoryController
+from quadruped_pympc.helpers.terrain_estimator import TerrainEstimator
+from quadruped_pympc.helpers.quadruped_utils import GaitType
+from gym_quadruped.utils.quadruped_utils import LegsAttr
+
+
+
+
+
+####
+    
+#------------
+class Controller:
+    #constructor
+    def __init__(self,
+                 env,
+                 gait_name=None,
+                 sim_dt=None,
+                 hip_height=None,
+                 legs_order=None,
+                 step_height=None,
+                 swing_position_gain_fb=None,
+                 swing_velocity_gain_fb=None,
+                 swing_generator=None,
+                 mpc_frequency= None,
+                 mpc_dt=None,
+                 controller_type=None,
+                 horizon=None,
+                 optimize_step_freq=None,
+                 step_freq_available=None,
+                 sampling_method=None,
+                 control_parametrization=None,
+                 sigma_cem_mppi=None,
+                 num_parallel_computations=None,
+                 num_sampling_iterations=None,
```

Comment:
As there are parameters that are specific to the type of MPC used under the hood, its best if the "MPC" specific parameters are not literally defined in the signature of the constructor, but rather are handled dynamically, something like:

```python
class SRBController:
   def __init__(
         arg_1,  # Arguments that all MPCs use. E.g., optimize_step_freq
         arg_2,
         ...
         mpc_type: str = 'nominal' # 'nominal' / 'input_rates'
         mpc_params: dict[str, Any] = None
      [...] 
      self.initialize_mpc(mpc_type, mpc_params)
   def set_controller(mpc_type, mpc_params):
       
       if mpc_type.lower() =='nominal':
           NominalClass(**mpc_params)  # This ensures each MPC class checks the appropriate parameters are passed, and if any irrelevant parameters are passed the class will through an unexpected argument Exception
   
```

Reply:
Ok  I will reformatt this. The init class of the controller will take as input the args and then two dictionaries. One with the simulation parameters and the other with the mpc paramters from the mpc.
It llooks something like this:
`    def __init__(self,
                 hip_height=None,
                 legs_order=None,        
                 inertia=None,
                 gait_params=None,
                 lift_off_positions=None,
                 simulation_parameters:dict=None,
                 mpc_parameters:dict=None,
                 ):`
Now each the mpc_prameter dict will be used in the set_contoller function and will call the specific parameters for each MPC. In order to do this I need to modify the classes for the gradient and sampling bassed controller. This I will do between wednesday and tomorrow. Other then that I fixed all the other issue I willp ush a new version so that you guys can check!